### PR TITLE
LIME-1273 - Adding new call to personInfo endpoint

### PR DESCRIPTION
**Pull request by ElliotMurphyGDS**
### What changed
 - Updated Root.js Controller to call new personInfo Endpoint and map returned values into sessionModel
 - made mapping of values to sessionModel conditional for compatibility with existing web route
 - Updated Config to include new personInfo endpoint path
 - Added tests for root controller

### CODE REVIEW

File: src/app/drivingLicence/controllers/root.js [hunk 1]
```diff
@@ -1,20 +1,74 @@
...
-      if (sharedClaims?.names?.length > 0) {
-        req.journeyModel.set("surname", sharedClaims.names[0]?.familyName);
-        req.journeyModel.set("givenNames", sharedClaims.names[0]?.givenNames);
-      }
+    const headers = {
+      session_id: req.session?.tokenId
+    };
+
+    //axios.post to new personInfo endpoint, put licence details into req.session.shared_claims shared claims
+    const personInfoApiResponse = await req.axios.post(
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.js:20]: For security can we wrap this in an if statement with an env var so we can merge without triggering new code yet

File: src/app/drivingLicence/controllers/root.js [hunk 2]
```diff
@@ -1,20 +1,74 @@
...
+      req.sessionModel.set(
+        "surname",
+        personInfoApiResponse.data.name[0].nameParts[0].FamilyName.value
+      );
+      req.sessionModel.set(
+        "dateOfBirth",
+        personInfoApiResponse.data.birthDate[0].value
+      );
     }
-    super.saveValues(req, res, next);
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.js:17]: Is this not still needed?

File: src/lib/config.js [hunk 3]
```diff
@@ -6,7 +6,8 @@ module.exports = {
     PATHS: {
       SESSION: "session",
       CHECK: "check-driving-licence",
-      AUTHORIZATION: "authorization"
+      AUTHORIZATION: "authorization",
+      PERSON_INFO: "person-info"
```
Comment (gds-jbentham) [inline src/lib/config.js:10]: Is this correct its different to whats on the ticket

File: src/app/drivingLicence/controllers/root.test.js [hunk 4]
```diff
@@ -1,84 +1,118 @@
...
     next = sinon.fake();
   });
   afterEach(() => sandbox.restore());
-
   it("should be an instance of BaseController", () => {
     expect(root).to.be.an.instanceof(BaseController);
   });
-
-  it("should retrieve shared_claims from session and store it in the journeyModel", async () => {
+  it("should retrieve shared_claims from session and store it in the sessionModel", async () => {
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.test.js:69]: Do we have a negative check for if any of these values are absent

File: src/lib/config.js (hunk 3 above)
Comment (ElliotMurphyGDS) [reply src/lib/config.js:10]: Endpoint on ticket was taken from a similar piece of work kiwi did so an internal decision not linked to any requirement, I opted to add the hyphen to keep inline with existing convention

File: src/app/drivingLicence/controllers/root.js (hunk 1 above)
Comment (ElliotMurphyGDS) [reply src/app/drivingLicence/controllers/root.js:20]: Yes this actually means I can remove feature toggle from corresponding API PR

File: src/app/drivingLicence/controllers/root.js [hunk 5]
```diff
@@ -1,19 +1,73 @@
...
-    const sharedClaims = req.session?.shared_claims;
+    req.sessionModel.reset();
+    req.sessionModel.set("isAuthSourceRoute", false);
 
-    if (sharedClaims) {
-      if (sharedClaims?.names?.length > 0) {
-        req.journeyModel.set("surname", sharedClaims.names[0]?.familyName);
-        req.journeyModel.set("givenNames", sharedClaims.names[0]?.givenNames);
-      }
+    let personInfoApiResponse = {};
```
Comment (smsgds) [inline src/app/drivingLicence/controllers/root.js:15]: The should be within the if for the feature flag.

File: src/app/drivingLicence/controllers/root.js [hunk 6]
```diff
@@ -1,19 +1,73 @@
...
+
+    if (process.env.AUTH_SOURCE_ENABLED === "true") {
+      //axios.post to new personInfo endpoint, put licence details into req.session.shared_claims shared claims
+      personInfoApiResponse = await req.axios.post(
+        `${BASE_URL}/${PERSON_INFO}`,
+        {},
+        { headers: headers }
+      );
+    }
+    if (personInfoApiResponse.data.drivingPermit) {
```
Comment (smsgds) [inline src/app/drivingLicence/controllers/root.js:29]: This should be in the above feature flag if.

It also needs to handle a failure of personInfoApiResponse not being returned and .data.drivingPermit not being present.

**Review by smsgds — CHANGES_REQUESTED** [review]

File: src/app/drivingLicence/controllers/root.js [hunk 7]
```diff
@@ -1,19 +1,73 @@
...
+    BASE_URL,
+    PATHS: { PERSON_INFO }
+  }
+} = require("../../../lib/config");
+
 class RootController extends BaseController {
   async saveValues(req, res, next) {
-    const sharedClaims = req.session?.shared_claims;
+    req.sessionModel.reset();
+    req.sessionModel.set("isAuthSourceRoute", false);
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.js:13]: Do we want to be resetting the session model we run the risk here of something failing and putting the user down a different route to the first one they were on if they navigate back

File: src/app/drivingLicence/controllers/root.js [hunk 8]
```diff
@@ -1,19 +1,73 @@
...
-      }
+    let personInfoApiResponse = {};
 
-      if (sharedClaims?.dateOfBirths?.length > 0) {
-        req.journeyModel.set("dateOfBirth", sharedClaims.dateOfBirths[0]);
+    const headers = {
+      session_id: req.session?.tokenId
+    };
+
+    if (process.env.AUTH_SOURCE_ENABLED === "true") {
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.js:21]: Should this be isAuthSourceRoute
if (process.env.AUTH_SOURCE_ENABLED && !isAuthSourceRoute)

That way we only run it once per journey

File: src/app/drivingLicence/controllers/root.js [hunk 9]
```diff
@@ -1,21 +1,108 @@
...
     super.saveValues(req, res, next);
   }
+
+  async checkForValidSharedClaimsData(req, personInfoApiResponse) {
+    if (
+      personInfoApiResponse.data?.drivingPermit[0]?.personalNumber &&
+      personInfoApiResponse.data?.drivingPermit[0]?.expiryDate &&
+      personInfoApiResponse.data?.drivingPermit[0]?.issueDate &&
+      personInfoApiResponse.data?.drivingPermit[0]?.issuedBy &&
+      personInfoApiResponse.data?.drivingPermit[0]?.issueNumber &&
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.js:78]: I think this needs to be more complicated. For instance we only expect an issue number if its DVLA

File: src/app/drivingLicence/controllers/root.js [hunk 10]
```diff
@@ -1,21 +1,108 @@
...
+        req.sessionModel.set(
+          "dateOfBirth",
+          personInfoApiResponse.data.birthDate[0].value
+        );
       }
     }
     super.saveValues(req, res, next);
   }
+
+  async checkForValidSharedClaimsData(req, personInfoApiResponse) {
```
Comment (gds-jbentham) [inline src/app/drivingLicence/controllers/root.js:72]: On naming can we make this return true or false and update the state in the calling method

**Review by smsgds — APPROVED** [review]

**Pull request merged**